DFN part generator

HBadertscher opened the pull request:
Hi,

I started working on a part generator for DFN packages according to JEDEC MO-229F. This is still work-in-progress, but I am already opening the pull request so we have a place to discuss certain issues:

- For many variants, e.g. V3020A (see Table 8 on sheet 8 of the JEDEC document), an exposed pad is allowed but not mandatory. E.g. in the case of the V3020A, an exposed pad up to 0.65mm x 2.25mm may exist, but it can also be 0.00 x 0.00, i.e. none. How do we handle this? Create a variant with (maximum size) exposed pad and without exposed pad? Or does it hurt to have only a variant with exposed pad?
- There are variants with different package heights: mostly 0.80mm vs 1.00mm, but there are also a few 0.65mm packages. Should I create separate packages for those, if the height is the only difference?
- Do you know of any specs for DFN footprints or should I just use some manufacturer's recommendation as a starting point and create reasonable defaults myself?
- What minimum silkscreen line width should the standard library have?

Thanks!

dbrgn commented:
Awesome!

As far as I know, DFN is just a special case of a QFN that doesn't have leads on two of the sides. See "Variants" at https://en.wikipedia.org/wiki/Quad_Flat_No-leads_package. Not sure if that information helps. See also IPC 7351 (http://pcbget.ru/Files/Standarts/IPC_7351.pdf) page 73 (labelled 63). Maybe you can add a follow-up PR that adds QFN support?

> Do you know of any specs for DFN footprints or should I just use some manufacturer's recommendation as a starting point and create reasonable defaults myself?

Maybe @rnestler @ubruhin or @fuerstr know more, since Sensirion uses DFN packages.

> Should I create separate packages for those, if the height is the only difference?

As far as I know from discussion with @ubruhin, two separate packages might be good, since they will be tied to 3D models later on.

> How do we handle this? Create a variant with (maximum size) exposed pad and without exposed pad? Or does it hurt to have only a variant with exposed pad?

I thought about using actual footprint variants for this, but that would require a device to be able to require a specific footprint variant, not just a package. Is that possible, @ubruhin? I don't think so.

We could always add the exposed pad, but then the stencils would have a hole there even if the DFN package does not have an exposed pad. That's not good either.

I'm unsure what the best solution is. We want to avoid combinatorial explosion, but every device should have a matching footprint. Maybe try to find out if there are 1-2 commonly used sizes and generate footprints for those.

As for naming, according to https://ohm.bu.edu/~pbohn/__Engineering_Reference/pcb_layout/pcbmatrix/IPC-7x51%20&%20PCBM%20Land%20Pattern%20Naming%20Convention.pdf the versions with heat sink would have a `_HS-xx` suffix:

![img](https://tmp.dbrgn.ch/screenshots/20181127151711-40vlmacw.png)

I don't know what the part after the suffix means. Maybe for a 2x2mm heat sink we could use `_HS-20X20`?

HBadertscher commented:
Adding QFN is definitely planned, though probably in a second step.

Using the `_HS` suffix for exposed pads would be an option, but as you say, it leads to a quite number of components. Right now I am already at 174 different DFN packages... Regarding different footprint versions: it could be an option to move the decision to use an exposed pad or not from the library designer to the PCB designer, wouldn't it? It might make sense, e.g. if I make a PCB for hand-soldering, I'm probably not going to use the exposed pad. 

About the land pattern condition: it also says that for QFN (and so probably also for DFN), one should always use Level B:

> Note: Ball Grid Array (BGA) and Quad Flat No-Lead (QFN) packages do not receive a suffix of ‘‘L,’’ ‘‘M’’ or ‘‘N’’ as they are only listed in the median (nominal) density level.

Do I understand the document you linked correctly that I could calculate the pad sizes using the maximum dimensions from JEDEC, and add the side/toe/heel-fillet from that IPC document to each side, e.g. for QFN for a pad of size 0.65mm x 0.5mm we would end up with pads of size 0.95mm x 0.4mm?

dbrgn commented:
By the way, right now you're saving the cache after every package you generate. Instead, you can save it only once at the end.

dbrgn commented:
And regarding heights: Maxim sometimes uses "Thin DFN" with 0.7-0.8mm height packages, for example for the MAX1848. Sensirion also uses that height (0.75mm) for their SHTC3 sensor in a DFN200X200X75-4 package. So maybe 0.75mm should be added too.

HBadertscher commented:
> By the way, right now you're saving the cache after every package you generate. Instead, you can save it only once at the end.

Oops, thanks, fixed.

> Maxim sometimes uses "Thin DFN" with 0.7-0.8mm height packages [...] So maybe 0.75mm should be added too.

Actually, those are the 0.80mm packages I am creating right now, which are according to JEDEC 0.70mm min / 0.75mm nominal / 0.80mm max. I am using the maximum value of 0.80mm.

> Sensirion also uses that height (0.75mm) for their SHTC3 sensor in a DFN200X200X75-4 package

It appears that the SHTC3 doesn't use a (JEDEC-) standard DFN package, but rather a Sensirion-specific DFN-style package. From what I see, JEDEC doesn't specify any 1.00mm pitch DFN devices, and no 2.0 x 2.0mm 4-pin-packages either. I can easily generate that package too, but I guess it would rather belong into a Sensirion-library and not in the Base library, right?

dbrgn commented:
> Actually, those are the 0.80mm packages I am creating right now, which are according to JEDEC 0.70mm min / 0.75mm nominal / 0.80mm max. I am using the maximum value of 0.80mm.

Ah, I see :slightly_smiling_face: I wonder if it would make sense to use the nominal height in the naming and the max height for the 3D model? What would you say, @ubruhin?

> I can easily generate that package too, but I guess it would rather belong into a Sensirion-library and not in the Base library, right?

Yes, you're probably right (unless there are other companies that use such a package).

HBadertscher commented:
I added non-JEDEC configs separately in a `3rd_party` output folder. We can choose to move those (only the SHTC-3 DFN package so far) wherever we like. 

> I wonder if it would make sense to use the nominal height in the naming and the max height for the 3D model?

I agree, fixed with the last commit.

So my next ToDos for this PR are the exposed pad and calculating good pad sizes and positions.

ubruhin commented:
I (finally) took a look at the generated packages - very nice work @HBadertscher! :smiley: 

So far the only thing I can say is that the text elements for `{{NAME}}` and `{{VALUE}}` are missing. And I had to run the script in a Docker container because it's not compatible with Python 3.5 from my system :cry: 

Btw, most Sensirion sensors should have separate packages anyway because they have a "hole" at the top and thus need a special 3D model. So those packages should be named accordingly to make clear they are intended only for Sensirion sensors.

dbrgn commented:
> Btw, most Sensirion sensors should have separate packages anyway because they have a "hole" at the top and thus need a special 3D model.

They could, although a 3D model without a hole would already be better than no 3D model at all.

The library format should support changing the package of a device after publication, as long as the version is incremented, right?

dbrgn commented:
> No, it's a breaking change when switching to a different package. The project library updater would become very complicated if we allow switching to different packages.

I think it would be fine if the project library updater would refuse to update packages with a changed package, and simply ask the user to manually remove and re-add the element, don't you think? This way there is no breakage, but the user still has the possibility of changing the package if desired.

(Use case: We start adding parts that reference generic packages, but later discover that not all of them have the same requirements. Just like the 3D-model-with-hole example. This *will* happen sooner or later.)

HBadertscher commented:
Thanks for the feedback. I added the `{{NAME}}` and `{{VALUE}}` now.

I need some help or verification: If i take e.g. the `V1515D` package (Table 11, sheet 9), and take the maximum lead length `L=0.65` and exposed pad size `E2=0.15`, then the clearance between the exposed pad and the lead is `(1.50 - 2*0.65 - 0.15) / 2 = 0.025`, which is way too low.  And it is the same for many other parts too. At the same time, they specify a clearance `K = 0.20` between pads and exposed pad, so I guess as a manufacturer you can choose to make either the pads or the exposed pad a bit larger or smaller...

If you want to check visually: right now the `generate_dfn.py` script generates all pads (including exposed pads) exactly at the maximum size specified. If you look at the footprints in the library editor you will see the tiny clearance.

Can you verify this or am I making a mistake? I plan on making both the pads and the exposed pad smaller such that the footprint also has a minimum clearance of 0.2 mm. Does this make sense? Should I make the clearance larger, e.g. 10mil (this is the default clearance in EAGLE)

ubruhin commented:
> Can you verify this or am I making a mistake?

I could verify it ;)

Maybe better use the nominal values instead of maximum values, and where the clearance then is still smaller than 0.2mm, make both the pads and the exposed pad smaller until the clearance is 0.2mm.

ubruhin commented:
Btw, [IPC-7351C seems to recommend a naming conventions for thermal pad sizes](http://ocipcdc.org/archive/What_is_New_in_IPC-7351C_03_11_2015.pdf):

![image](https://user-images.githubusercontent.com/5374821/49764721-889a2280-fcd0-11e8-8c9b-c303681ffe03.png)

HBadertscher commented:
I think this is getting more and more ready to merge:

1. Added the `T` suffix for the exposed pad sizes.
2. I use the naming `DFN{pitch}P...`, as with standard IPC-rules for the DFN the pitch is not in the name, leading to duplicate names. I thus added the pitch in the same way as it is used for the closely related QFN packages.
3. As suggested, I now use the nominal pad and exposed pad sizes. If the clearance is smaller than 0.20mm, both the pads and the exposed pad are shrunk until the clearance is 0.20mm. Except in cases where the exposed pad would get smaller than the minimum trace width of 0.10mm, in which case only the pads are shrunk. I think this is an elegant solution.
4. Added the exact lead dimensions of both the pads and the exposed pads to the documentation layer.
5. To calculate the pad size, I use the recommended toe/heel size from [IPC-7351C (slide 26)](http://ocipcdc.org/archive/What_is_New_in_IPC-7351C_03_11_2015.pdf). I only add that to the "toe" of the pad, and not the "heel" as not to have problems with the clearance. For the hand-soldering versions, I add 0.3mm to the toe of the pad.
6. Open issue: In some cases, the footprints only differ by the size of the pads/leads, which leads to duplicate names according to IPC-rules. I think I somewhere found that the suffix `P{lead length}` can be used for that, but I cannot find the source of that anymore. So right now, we have for example:
    - `DFN50P300X200X95-10P30T220X60`
    - `DFN50P300X200X95-10P55T220X60 `
for two footprints where the difference is the lead length (0.30mm vs 0.55mm). Is that ok? Do you know of an official way? The `P` seems wrong as it is already used for the pitch...

I'm happy about reviews and comments @dbrgn @ubruhin  :)

dbrgn commented on dfn_configs.py:
```diff
@@ -0,0 +1,259 @@
+
+THIRD_CONFIGS = [
+        # Sensirion SHTC-3
+        DfnConfig(2.0, 2.0, 1.0, 4, 0.75, 0.80, 0.35, 1.60, 0.70, 'SHTC-3', lead_width=0.35),
```
The `SHTC3` doesn't have a dash in the name AFAIK.

Also, what about allowing a name prefix in the DfnConfig? For example, this one could be called "Sensirion DFNxxxxx". You could also add the manufacturer name ("sensirion") to the keywords.

HBadertscher replied:
Hm, the package will go into the Sensirion library, so it should be clear that it is a Sensirion part. Shouldn't that be enough?

I added the manufacturer name to the keywords as suggested.

HBadertscher replied:
Oh I wasn't aware of that, sorry. Fixed in 
00a88d4

HBadertscher commented on dfn_configs.py:
```diff
@@ -0,0 +1,259 @@
+        DfnConfig(1.5, 1.5, 0.5, 4, 0.95, 1.00, 0.55, 0.70, 0.10, 'V1515D,VBBD'),
+        DfnConfig(1.5, 1.5, 0.5, 4, 0.75, 0.80, 0.55, 0.70, 0.10, 'W1515D,WBBD'),
+        # Square, 2.0 x 2.0
+        DfnConfig(2.0, 2.0, 0.65, 6, 0.95, 1.00, 0.30, 1.58, 0.65, 'V2020C,VCCC'),    # no nominal exp_pad
```
This is a different thing: if the `no_exp` flag is true (which is the default), then we create two variants: one with exposed pad, and one without exposed pad. If `no_exp` is false, then we only create a variant *with* exposed pad.

The comment indicates that JEDEC only defines a minimal and maximal exposed pad size, but no nominal values. In such cases I tried to take sensible defaults, but left a comment saying that these are not the nominal values of the JEDEC standard.

HBadertscher commented on dfn_configs.py:
```diff
@@ -0,0 +1,259 @@
+
+"""
+
+# Maximal lead width as a function of pitch, Table 4
```
Yes you are right, fixed in c424029

HBadertscher commented on dfn_configs.py:
```diff
@@ -0,0 +1,259 @@
+        try:
+            self.toe_heel = LEAD_TOE_HEEL[pitch]
+        except KeyError:
+            raise NotImplementedError(f"No toe/heel length for pitch {pitch}")
```
Ok sure, fixed in c424029

rnestler commented:
@HBadertscher To squash the `--fixup` commit you can use: [`git rebase --autosquash`](https://robots.thoughtbot.com/autosquashing-git-commits)

HBadertscher commented:
I squashed the fixup commits as suggested by @rnestler and I think the other commits do make sense the way they are. Is this ok @dbrgn ?

dbrgn commented:
Yep, thanks @HBadertscher!
